File: bindings/python/perf/multi/perf_multi_pubsub_server.py
import os
import sys
import threading
import time

import zlink

from perf_multi_common import (
    TOPIC,
    apply_multi_socket_options,
    benchmark_endpoint,
    benchmark_run_id,
    new_payload,
    parse_server_args,
    publish_nonblocking,
    stamp_payload,
)


def _stdin_stop(stop_event):
    for line in sys.stdin:
        if line.strip().upper() in {"STOP", "QUIT"}:
            stop_event.set()
            return


def main(argv=None):
    args = parse_server_args(argv or sys.argv[1:])
    run_id = benchmark_run_id()
    stop_event = threading.Event()
    start_event = threading.Event()
    endpoint = benchmark_endpoint(args.transport, "multi-pubsub")
    payload = new_payload(args.msg_size)
    active_duration_s = max(
        1.0,
        float(os.environ.get("PERF_MULTI_DURATION_SECONDS", "5")),
    )

    def read_commands():
        for line in sys.stdin:
            text = line.strip().upper()
            if text.startswith("START,"):
                start_event.set()
            elif text in {"STOP", "QUIT"}:
                stop_event.set()
                return

    threading.Thread(target=read_commands, daemon=True).start()

    with zlink.Context() as ctx:
        with zlink.PubSocket(ctx) as publisher:
            apply_multi_socket_options(publisher)
            publisher.bind(endpoint)
            print(f"READY,{endpoint}", flush=True)
            active_deadline = None
            while not stop_event.is_set():
                if not start_event.is_set():
                    stop_event.wait(0.01)
                    continue
                if active_deadline is None:
                    active_deadline = time.perf_counter() + active_duration_s
                if time.perf_counter() >= active_deadline:
                    stop_event.wait(0.01)
                    continue
                sent_any = False
                while time.perf_counter() < active_deadline and not stop_event.is_set():
                    sent = publish_nonblocking(
                        publisher,
                        TOPIC,
                        stamp_payload(payload, phase=1, run_id=run_id),
                    )
                    if not sent:
                        break
                    sent_any = True
                if not sent_any:
                    stop_event.wait(0.001)


if __name__ == "__main__":
    main()

File: bindings/node/perf/multi/perf_multi_pubsub_server.ts
// SPDX-License-Identifier: MPL-2.0

'use strict';

const readline = require('node:readline');
const zlink = require('@zlink-systems/zlink');
const {
  createPayload,
  createRunId,
  stampPayload
} = require('../common/perf_metrics');
const { configureTlsServer } = require('../common/perf_tls');
const { parseMultiArgs } = require('./perf_multi_common');
const {
  POLLOUT,
  applyAutoHwmMsgUnit,
  applyContextPolicy,
  applySocketPolicy,
  emitMultiSocketHwmDetail,
  pollEvents,
  pollEventHas,
  trySocketPublish
} = require('./perf_multi_runtime');

async function main() {
  const options = parseMultiArgs(process.argv.slice(2));
  const ctx = new zlink.Context();
  applyContextPolicy(ctx, 'server', 'MULTI_PUBSUB');
  const pub = new zlink.PubSocket(ctx);
  const poller = new zlink.Poller();
  const payload = createPayload(options.msgSize);
  let rl = null;

  try {
    applySocketPolicy(pub, {
      noDrop: Number(process.env.PERF_MULTI_PUBSUB_XPUB_NODROP ?? 1) !== 0
    });
    configureTlsServer(pub, options.transport);
    pub.bind(options.endpoint);
    applyAutoHwmMsgUnit(pub, options.msgSize);
    ctx.recalculateAutoHwm();
    emitMultiSocketHwmDetail(pub, 'endpoint', options.transport, options.msgSize);
    poller.add(pub, pollEvents(POLLOUT));
    console.log(`READY,${options.endpoint}`);

    rl = readline.createInterface({ input: process.stdin, crlfDelay: Infinity });
    for await (const line of rl) {
      if (line !== `START,${options.msgSize}`) {
        if (line === 'STOP' || line === 'QUIT') {
          break;
        }
        continue;
      }

      const runId = createRunId(1);
      const activeStopNs = process.hrtime.bigint() + BigInt(Math.floor(options.duration * 1_000_000_000));
      let seq = 1n;
      let pending = false;
      while (process.hrtime.bigint() < activeStopNs) {
        if (!pending) {
          stampPayload(payload, { phase: 1, runId, msgSize: options.msgSize, seq });
          if (trySocketPublish(pub, 'perf.topic', payload)) {
            seq += 1n;
            continue;
          }
          pending = true;
        }

        // PERF_MULTI_TEST_POLICY § 1.3.1: signal-driven `-1` wait.
        const ready = poller.wait(-1);
        if (!ready || !pollEventHas(ready, POLLOUT)) {
          continue;
        }
        pending = false;
        if (process.hrtime.bigint() >= activeStopNs) {
          break;
        }
      }
      stampPayload(payload, { phase: 2, runId, msgSize: options.msgSize, seq });
      for (let sent = 0; sent < Math.max(options.clients * 16, options.clients);) {
        if (trySocketPublish(pub, 'perf.topic', payload)) {
          sent += 1;
          continue;
        }
        const ready = poller.wait(-1);
        if (!ready || !pollEventHas(ready, POLLOUT)) {
          continue;
        }
      }
      break;
    }
  } finally {
    rl?.close();
    poller.close();
    pub.close();
    ctx.close();
  }
}

main().catch((error) => {
  console.error(error);
  process.exitCode = 1;
});
